File: 2015/07/solution.js
let circuit = {}
let signals = {}

const evaluate = (definition) => {
  let match

  if (match = definition.match(/^(\w+)$/)) {
    return gate(match[1])
  }

  if (match = definition.match(/^NOT (\w+)$/)) {
    return ~ gate(match[1])
  }

  if (match = definition.match(/^(\w+) AND (\w+)$/)) {
    return gate(match[1]) & gate(match[2])
  }

  if (match = definition.match(/^(\w+) OR (\w+)$/)) {
    return gate(match[1]) | gate(match[2])
  }

  if (match = definition.match(/^(\w+) RSHIFT (\w+)$/)) {
    return gate(match[1]) >> gate(match[2])
  }

  if (match = definition.match(/^(\w+) LSHIFT (\w+)$/)) {
    return gate(match[1]) << gate(match[2])
  }

  throw `Unexpected definition: ${definition}`
}

const addGateToCircuit = (line) => {
  const [, definition, outGate] = line.match(/^(.*) -> (\w+)$/)
  circuit[outGate] = definition
}

const gate = (reference) => {
  const val = parseInt(reference)
  if (`${val}` === reference) return val
  if (typeof signals[reference] === 'undefined') {
    signals[reference] = evaluate(circuit[reference])
  }
  return signals[reference]
}

const input = require('fs').createReadStream('input.txt')
const reader = require('readline').createInterface({ input })

reader.on('line', addGateToCircuit)
reader.on('close', () => {
  const initialValue = gate('a')
  console.log(`Gate a initially receives signal of ${initialValue}`)
  signals = { b: initialValue } 
  console.log(`Overridding b to ${initialValue} and resetting the circuit, a receives ${gate('a')}`)
})

File: 2015/07/solution.rb
class Circuit

  def initialize
    @circuit = {}
    @signals = {}
  end

  def evaluate(definition)
    if result = /^(\w+)$/.match(definition)
      return gate(result[1])
    end

    if result = /^NOT (\w+)$/.match(definition)
      return ~ gate(result[1])
    end

    if result = /^(\w+) AND (\w+)$/.match(definition)
      return gate(result[1]) & gate(result[2])
    end

    if result = /^(\w+) OR (\w+)$/.match(definition)
      return gate(result[1]) | gate(result[2])
    end

    if result = /^(\w+) RSHIFT (\w+)$/.match(definition)
      return gate(result[1]) >> gate(result[2])
    end

    if result = /^(\w+) LSHIFT (\w+)$/.match(definition)
      return gate(result[1]) << gate(result[2])
    end
  end

  def define_gate(line)
    definition, out_gate = /^(.*) -> (\w+)$/.match(line)[1..2]
    @circuit[out_gate] = definition
  end

  def gate(reference)
    val = reference.to_i
    return val if val.to_s == reference
    @signals[reference] ||= evaluate(@circuit[reference])
  end

  def reset(signals)
    @signals = signals
  end

end

circuit = Circuit.new

File.open('input.txt').each_line do |line|
  circuit.define_gate line
end

puts 'Part 1', circuit.gate('a')
circuit.reset({ 'b' => circuit.gate('a') })
puts 'Part 2', circuit.gate('a')
